File: ShellingLaptopApp/app/src/main/java/com/shellinglaptop/adapter/LaptopAdapter.java
package com.shellinglaptop.adapter;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.shellinglaptop.R;
import com.shellinglaptop.databinding.ItemLaptopBinding;
import com.shellinglaptop.model.Laptop;
import com.shellinglaptop.utils.PriceUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public class LaptopAdapter extends RecyclerView.Adapter<LaptopAdapter.LaptopViewHolder>{

    private List<Laptop> laptops;
    private RecyclerViewLaptopClickListener recyclerViewLaptopClickListener;
    private List<Laptop> searchLaptop;

    public void setRecyclerViewLaptopClickListener(RecyclerViewLaptopClickListener recyclerViewLaptopClickListener) {
        this.recyclerViewLaptopClickListener = recyclerViewLaptopClickListener;
    }

    public LaptopAdapter(List<Laptop> laptops) {
        searchLaptop = new ArrayList<>();
        this.laptops = laptops;
    }

    public void setLaptops(List<Laptop> laptops){
        this.laptops = laptops;
        searchLaptop.addAll(laptops);
        notifyDataSetChanged();
    }

    @Override
    public LaptopViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new LaptopViewHolder(
            DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), R.layout.item_laptop, parent, false),
            recyclerViewLaptopClickListener
        );
    }

    @Override
    public void onBindViewHolder(@NonNull LaptopViewHolder holder, int position) {
        laptops.get(position).setPriceStr(laptops.get(position).getPrice().toString());
        String setupMoney = PriceUtils.setupPrice(laptops.get(position).getPriceStr());
        laptops.get(position).setPriceStr(setupMoney +" VND");
        holder.itemLaptopBinding.setLaptop(laptops.get(position));
        holder.itemLaptopBinding.getRoot().setOnClickListener(v ->
                holder.recyclerViewLaptopClickListener.RecyclerViewLaptopItemClick(laptops.get(position)));
    }

    @Override
    public int getItemCount() {
        return laptops.size();
    }

    public void filter(String text){
        String myText = text.toLowerCase(Locale.getDefault());
        laptops.clear();
        if(myText.isEmpty()){
            laptops.addAll(searchLaptop);
        }else {
            for(Laptop laptop : searchLaptop){
                if(laptop.getName().toLowerCase(Locale.getDefault()).contains(myText)){
                    laptops.add(laptop);
                }
            }
        }

        notifyDataSetChanged();
    }

    class LaptopViewHolder extends RecyclerView.ViewHolder {

        private ItemLaptopBinding itemLaptopBinding;
        private RecyclerViewLaptopClickListener recyclerViewLaptopClickListener;

        public LaptopViewHolder(ItemLaptopBinding itemLaptopBinding, RecyclerViewLaptopClickListener recyclerViewLaptopClickListener) {
            super(itemLaptopBinding.getRoot());
            this.itemLaptopBinding = itemLaptopBinding;
            this.recyclerViewLaptopClickListener = recyclerViewLaptopClickListener;
        }
    }

    public interface RecyclerViewLaptopClickListener{
        void RecyclerViewLaptopItemClick(Laptop laptop);
    }
}

File: ShellingLaptopApp/app/src/main/java/com/example/shellinglaptopapp/ui/laptops/LaptopAdapter.kt
package com.example.shellinglaptopapp.ui.laptops

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.example.shellinglaptopapp.R
import com.example.shellinglaptopapp.databinding.ItemLaptopBinding
import com.example.shellinglaptopapp.data.model.Laptop
import java.util.*
import kotlin.collections.ArrayList

class LaptopAdapter(
    private val laptops: ArrayList<Laptop>,
    private val listener: RecyclerViewLaptopClickListener
): RecyclerView.Adapter<LaptopAdapter.LaptopViewHolder>() {

    private var searchLaptops: ArrayList<Laptop> = ArrayList()
    init {
        searchLaptops.addAll(laptops)
    }

    class LaptopViewHolder(
        val itemLaptopBinding: ItemLaptopBinding
    ) : RecyclerView.ViewHolder(itemLaptopBinding.root)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LaptopViewHolder =
        LaptopViewHolder(
            DataBindingUtil.inflate(
                LayoutInflater.from(parent.context),
                R.layout.item_laptop,
                parent, false)
        )

    override fun onBindViewHolder(holder: LaptopViewHolder, position: Int) {
        var reversePrice = reverseString(laptops[position].price.toString())
        var newPriceStr = "" //add "."
        var count = 0
        for(element in reversePrice){

            if(count == 3){
                newPriceStr = newPriceStr.plus(".")
                count = 0
            }
            count++
            newPriceStr = newPriceStr.plus(element)
        }

        var reverseNewTotalMoneyStr = reverseString(newPriceStr)
        laptops[position].priceStr = "$reverseNewTotalMoneyStr VNĐ"
        holder.itemLaptopBinding.laptop = laptops[position]

        holder.itemLaptopBinding.root.setOnClickListener {
            listener.onRecyclerViewItemClick(laptops[position])
        }
    }

    private fun reverseString(str: String): String{
        var position = str.length-1
        var strReverse = ""
        for(e in str){
            strReverse = strReverse.plus(str[position])
            position--
        }

        return strReverse
    }

    override fun getItemCount(): Int = laptops.size

    fun filter(text: String){
        var myText = text.toLowerCase(Locale.getDefault())
        laptops.clear()
        if(myText.isEmpty()) {
            laptops.addAll((searchLaptops))
        }else{
            for(laptop in searchLaptops){
                if(laptop.name!!.toLowerCase(Locale.getDefault()).contains(myText)){
                    laptops.add(laptop)
                }
            }
        }

        notifyDataSetChanged()
    }
}